File: recaller/policy_engine/engine.py
"""RECALLER — policy engine.

Consumes deterministic credit metrics plus reconciliation and confidence
state, walks the configured rule set, and returns a verdict with reason
codes. The engine holds no thresholds of its own: every number it compares
against comes out of the policy document.
"""

from typing import Any, Dict, List, Optional, Tuple, Union
from ..core.constants import DECISIONS
from ..core.hash import hash_value, short_hash


class OUTCOME:
    PASS = "PASS"
    REFER = "REFER"
    FAIL = "FAIL"
    NOT_APPLICABLE = "NOT_APPLICABLE"


Sequence_Band = Union[List[float], Tuple[float, ...]]


def in_band(value: float, band: Optional[Sequence_Band]) -> bool:
    if not band or len(band) < 2:
        return False
    lo, hi = band[0], band[1]
    return lo <= value <= hi



def fmt(v: Any) -> str:
    if not isinstance(v, (int, float)):
        return str(v)
    if isinstance(v, float) and v.is_integer():
        return str(int(v))
    if isinstance(v, int):
        return str(v)
    # Format up to 4 decimal places without trailing zeroes
    formatted = f"{float(v):.4f}".rstrip("0").rstrip(".")
    return formatted


def evaluate_rule(rule: Dict[str, Any], ctx: Dict[str, Any]) -> Dict[str, Any]:
    """Evaluate one rule against the metric bag."""
    metrics = ctx.get("metrics", {})
    segment = ctx.get("segment", {})
    metric_key = rule.get("metric")
    actual = metrics.get(metric_key)

    shell = {
        "code": rule.get("code"),
        "label": rule.get("label"),
        "metric": metric_key,
        "severity": rule.get("severity"),
        "rationale": rule.get("rationale"),
        "operator": rule.get("operator"),
    }

    if actual is None:
        return {
            **shell,
            "outcome": OUTCOME.NOT_APPLICABLE,
            "actual": None,
            "threshold": rule.get("threshold"),
            "reason": None,
            "detail": "Metric unavailable for this application.",
        }

    op = rule.get("operator")
    threshold = rule.get("threshold")
    refer_band = rule.get("refer_band")

    if op == "lte":
        if actual <= threshold:
            outcome = OUTCOME.PASS
        elif in_band(actual, refer_band):
            outcome = OUTCOME.REFER
        else:
            outcome = OUTCOME.FAIL
        sign = "≤" if actual <= threshold else ">"
        detail = f"{fmt(actual)} {sign} {fmt(threshold)}"

    elif op == "gte":
        if actual >= threshold:
            outcome = OUTCOME.PASS
        elif in_band(actual, refer_band):
            outcome = OUTCOME.REFER
        else:
            outcome = OUTCOME.FAIL
        sign = "≥" if actual >= threshold else "<"
        detail = f"{fmt(actual)} {sign} {fmt(threshold)}"

    elif op == "within_segment_ticket":
        min_ticket = segment.get("min_ticket", 0)
        max_ticket = segment.get("max_ticket", 0)
        threshold = {"min": min_ticket, "max": max_ticket}
        outcome = OUTCOME.PASS if min_ticket <= actual <= max_ticket else OUTCOME.FAIL
        detail = f"{fmt(actual)} within {fmt(min_ticket)}–{fmt(max_ticket)}"

    elif op == "within_segment_tenure":
        t_min = segment.get("tenure_months", {}).get("min", 0)
        t_max = segment.get("tenure_months", {}).get("max", 0)
        threshold = {"min": t_min, "max": t_max}
        outcome = OUTCOME.PASS if t_min <= actual <= t_max else OUTCOME.FAIL
        detail = f"{actual}m within {t_min}–{t_max}m"

    else:
        raise ValueError(f"Unsupported policy operator: {op}")

    if outcome == OUTCOME.PASS:
        reason = rule.get("pass_reason")
    elif outcome == OUTCOME.REFER:
        reason = rule.get("refer_reason")
    else:
        reason = rule.get("reject_reason") or rule.get("refer_reason")

    return {
        **shell,
        "outcome": outcome,
        "actual": actual,
        "threshold": threshold,
        "reason": reason,
        "detail": detail,
    }


def evaluate_policy(args: Dict[str, Any]) -> Dict[str, Any]:
    """Run the full policy against credit metrics and reconciliation state."""
    metrics = args.get("metrics", {})
    reconciliation = args.get("reconciliation") or {}
    unresolved_low_confidence = args.get("unresolvedLowConfidence", 0)
    loan_request = args.get("loanRequest", {})
    policy = args.get("policy", {})

    segment_code = loan_request.get("segment")
    segment = policy.get("segments", {}).get(segment_code)
    if not segment:
        raise ValueError(f"Unknown asset segment: {segment_code}")

    bag = {
        **metrics,
        "blocking_finding_count": reconciliation.get("blocking", 0),
        "advisory_finding_count": reconciliation.get("advisory", 0),
        "unresolved_low_confidence_count": unresolved_low_confidence,
    }

    results = [evaluate_rule(rule, {"metrics": bag, "segment": segment}) for rule in policy.get("rules", [])]

    hard_fails = [
        r
        for r in results
        if r["outcome"] == OUTCOME.FAIL
        and r["severity"] == "BLOCKING"
        and r.get("reason")
        and r["reason"].startswith("R")
    ]
    advisory_fails = [
        r for r in results if r["outcome"] == OUTCOME.FAIL and r["severity"] == "ADVISORY"
    ]
    refers = [r for r in results if r["outcome"] == OUTCOME.REFER]

    reason_codes = []

    if hard_fails:
        decision = DECISIONS.REJECT
        for r in hard_fails:
            if r.get("reason"):
                reason_codes.append(r["reason"])
        for r in refers:
            if r.get("reason"):
                reason_codes.append(r["reason"])
    elif refers or advisory_fails or unresolved_low_confidence > 0:
        decision = DECISIONS.REFER
        for r in refers:
            if r.get("reason"):
                reason_codes.append(r["reason"])
        for r in advisory_fails:
            if r.get("reason"):
                reason_codes.append(r["reason"])
        if unresolved_low_confidence > 0 and "F01" not in reason_codes:
            reason_codes.append("F01")
    else:
        decision = DECISIONS.APPROVE
        for r in results:
            if r["outcome"] == OUTCOME.PASS and r.get("reason"):
                reason_codes.append(r["reason"])

    unique_codes = sorted(set(reason_codes))
    policy_reason_defs = policy.get("reason_codes", {})

    return {
        "decision": decision,
        "reason_codes": [
            {"code": code, "text": policy_reason_defs.get(code, code)}
            for code in unique_codes
        ],
        "rules": results,
        "summary": {
            "passed": sum(1 for r in results if r["outcome"] == OUTCOME.PASS),
            "referred": len(refers),
            "failed": sum(1 for r in results if r["outcome"] == OUTCOME.FAIL),
            "not_applicable": sum(1 for r in results if r["outcome"] == OUTCOME.NOT_APPLICABLE),
            "total": len(results),
        },
        "policy_version": policy.get("version"),
        "policy_id": policy.get("policy_id"),
        "policy_effective_date": policy.get("effective_date"),
        "policy_hash": short_hash(policy),
        "evaluated_hash": hash_value({"bag": bag, "version": policy.get("version")}),
    }


def decisive_rules(evaluation: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Rules the UI should foreground for a given decision."""
    if evaluation.get("decision") == DECISIONS.APPROVE:
        return [r for r in evaluation.get("rules", []) if r["outcome"] == OUTCOME.PASS]
    return [
        r
        for r in evaluation.get("rules", [])
        if r["outcome"] in (OUTCOME.FAIL, OUTCOME.REFER)
    ]

File: packages/policy-engine/src/engine.js
/**
 * RECALLER — policy engine.
 *
 * JavaScript port of recaller/policy_engine/engine.py.
 * Consumes deterministic credit metrics plus reconciliation and confidence
 * state, walks the configured rule set, and returns a verdict with reason
 * codes. The engine holds no thresholds of its own: every number it compares
 * against comes out of the policy document.
 */

import { DECISIONS } from '@core/constants.js'
import { hashValue, shortHash } from '@core/hash.js'

const OUTCOME = Object.freeze({
  PASS: 'PASS',
  REFER: 'REFER',
  FAIL: 'FAIL',
  NOT_APPLICABLE: 'NOT_APPLICABLE',
})

function inBand(value, band) {
  if (!band || band.length < 2) return false
  return band[0] <= value && value <= band[1]
}

function fmt(v) {
  if (typeof v !== 'number') return String(v)
  if (Number.isInteger(v)) return String(v)
  return parseFloat(v.toFixed(4)).toString()
}

function evaluateRule(rule, ctx) {
  const metrics = ctx.metrics || {}
  const segment = ctx.segment || {}
  const metricKey = rule.metric
  const actual = metrics[metricKey]

  const shell = {
    code: rule.code,
    label: rule.label,
    metric: metricKey,
    severity: rule.severity,
    rationale: rule.rationale,
    operator: rule.operator,
  }

  if (actual == null) {
    return {
      ...shell,
      outcome: OUTCOME.NOT_APPLICABLE,
      actual: null,
      threshold: rule.threshold,
      reason: null,
      detail: 'Metric unavailable for this application.',
    }
  }

  const op = rule.operator
  let threshold = rule.threshold
  const referBand = rule.refer_band
  let outcome, detail

  if (op === 'lte') {
    if (actual <= threshold) outcome = OUTCOME.PASS
    else if (inBand(actual, referBand)) outcome = OUTCOME.REFER
    else outcome = OUTCOME.FAIL
    const sign = actual <= threshold ? '≤' : '>'
    detail = `${fmt(actual)} ${sign} ${fmt(threshold)}`
  } else if (op === 'gte') {
    if (actual >= threshold) outcome = OUTCOME.PASS
    else if (inBand(actual, referBand)) outcome = OUTCOME.REFER
    else outcome = OUTCOME.FAIL
    const sign = actual >= threshold ? '≥' : '<'
    detail = `${fmt(actual)} ${sign} ${fmt(threshold)}`
  } else if (op === 'within_segment_ticket') {
    const minTicket = segment.min_ticket || 0
    const maxTicket = segment.max_ticket || 0
    threshold = { min: minTicket, max: maxTicket }
    outcome = minTicket <= actual && actual <= maxTicket ? OUTCOME.PASS : OUTCOME.FAIL
    detail = `${fmt(actual)} within ${fmt(minTicket)}–${fmt(maxTicket)}`
  } else if (op === 'within_segment_tenure') {
    const tMin = (segment.tenure_months || {}).min || 0
    const tMax = (segment.tenure_months || {}).max || 0
    threshold = { min: tMin, max: tMax }
    outcome = tMin <= actual && actual <= tMax ? OUTCOME.PASS : OUTCOME.FAIL
    detail = `${actual}m within ${tMin}–${tMax}m`
  } else {
    throw new Error(`Unsupported policy operator: ${op}`)
  }

  let reason
  if (outcome === OUTCOME.PASS) reason = rule.pass_reason
  else if (outcome === OUTCOME.REFER) reason = rule.refer_reason
  else reason = rule.reject_reason || rule.refer_reason

  return { ...shell, outcome, actual, threshold, reason, detail }
}

/**
 * Run the full policy against credit metrics and reconciliation state.
 */
export function evaluatePolicy({ metrics, reconciliation, unresolvedLowConfidence = 0, loanRequest, policy }) {
  const segmentCode = loanRequest.segment
  const segment = (policy.segments || {})[segmentCode]
  if (!segment) throw new Error(`Unknown asset segment: ${segmentCode}`)

  const bag = {
    ...metrics,
    blocking_finding_count: (reconciliation || {}).blocking || 0,
    advisory_finding_count: (reconciliation || {}).advisory || 0,
    unresolved_low_confidence_count: unresolvedLowConfidence,
  }

  const results = (policy.rules || []).map((rule) => evaluateRule(rule, { metrics: bag, segment }))

  const hardFails = results.filter(
    (r) => r.outcome === OUTCOME.FAIL && r.severity === 'BLOCKING' && r.reason && r.reason.startsWith('R')
  )
  const advisoryFails = results.filter(
    (r) => r.outcome === OUTCOME.FAIL && r.severity === 'ADVISORY'
  )
  const refers = results.filter((r) => r.outcome === OUTCOME.REFER)

  const reasonCodes = []

  let decision
  if (hardFails.length > 0) {
    decision = DECISIONS.REJECT
    for (const r of hardFails) if (r.reason) reasonCodes.push(r.reason)
    for (const r of refers) if (r.reason) reasonCodes.push(r.reason)
  } else if (refers.length > 0 || advisoryFails.length > 0 || unresolvedLowConfidence > 0) {
    decision = DECISIONS.REFER
    for (const r of refers) if (r.reason) reasonCodes.push(r.reason)
    for (const r of advisoryFails) if (r.reason) reasonCodes.push(r.reason)
    if (unresolvedLowConfidence > 0 && !reasonCodes.includes('F01')) reasonCodes.push('F01')
  } else {
    decision = DECISIONS.APPROVE
    for (const r of results) if (r.outcome === OUTCOME.PASS && r.reason) reasonCodes.push(r.reason)
  }

  const uniqueCodes = [...new Set(reasonCodes)].sort()
  const policyReasonDefs = policy.reason_codes || {}

  return {
    decision,
    reason_codes: uniqueCodes.map((code) => ({ code, text: policyReasonDefs[code] || code })),
    rules: results,
    summary: {
      passed: results.filter((r) => r.outcome === OUTCOME.PASS).length,
      referred: refers.length,
      failed: results.filter((r) => r.outcome === OUTCOME.FAIL).length,
      not_applicable: results.filter((r) => r.outcome === OUTCOME.NOT_APPLICABLE).length,
      total: results.length,
    },
    policy_version: policy.version,
    policy_id: policy.policy_id,
    policy_effective_date: policy.effective_date,
    policy_hash: shortHash(policy),
    evaluated_hash: hashValue({ bag, version: policy.version }),
  }
}

export { OUTCOME }
